Reconcile SKU in 3scale

makslion opened the pull request:
# Description
https://issues.redhat.com/browse/MGDAPI-1225
Implementation of the dynamic SKU configuration for 3scale. 

## Verification steps
### Install the operator
By running the following commands

```
INSTALLATION_TYPE=managed-api make cluster/prepare/local
INSTALLATION_TYPE=managed-api IN_PROW=true make deploy/integreatly-rhmi-cr.yml
INSTALLATION_TYPE=managed-api make code/run
```
During the first installation verify that the operator has been provided with a secret 1

`oc get rhmi rhoam -n redhat-rhoam-operator -o yaml | yq - r 'status.toSKU'` -> 1

Wait for the installation to fully complete. You can verify this by
`oc get rhmi rhoam -n redhat-rhoam-operator -o yaml | yq - r 'status.toSKU'` -> nil

### Verify that the 3scale has expected values

#### Memory and cpu 
Apicast production: 
```oc get apimanager 3scale -n redhat-rhoam-3scale -o yaml | yq - r 'spec.apicast.productionSpec.resources'```

Output:
```
limits:
  cpu: 100m
  memory: 270Mi
requests:
  cpu: 90m
  memory: 250Mi
```

Backend listeners: 
```
oc get apimanager 3scale -n redhat-rhoam-3scale -o yaml | yq - r 'spec.backend.listenerSpec.resources'
```
Output: 
```
limits:
  cpu: 110m
  memory: 500Mi
requests:
  cpu: 100m
  memory: 450Mi
```

Backend workers: 
```
oc get apimanager 3scale -n redhat-rhoam-3scale -o yaml | yq - r 'spec.backend.workerSpec.resources'
```

Output: 
```
limits:
  cpu: 40m
  memory: 60Mi
requests:
  cpu: 30m
  memory: 55Mi
```

#### Replicas 
Apicast production:

`oc get apimanager 3scale -n redhat-rhoam-3scale -o yaml | yq - r 'spec.apicast.productionSpec.replicas'`-> 1

Backend listeners: 

`oc get apimanager 3scale -n redhat-rhoam-3scale -o yaml | yq - r 'spec.backend.listenerSpec.replicas' ` -> 1

Backend workers: 

`oc get apimanager 3scale -n redhat-rhoam-3scale -o yaml | yq - r 'spec.backend.workerSpec.replicas'` -> 1


### Verify that changing the sku will update configuration
Run `INSTALLATION_TYPE=managed-api QUOTA=20 make cluster/prepare/sku`

which will update the secret with a sku value of 20 on the next reconcile

`oc get rhmi rhoam -n redhat-rhoam-operator -o yaml | yq - r 'status.toSKU'` -> 20

Wait for the reconcile to complete

`oc get rhmi rhoam -n redhat-rhoam-operator -o yaml | yq - r 'status.toSKU'` -> nil
`oc get rhmi rhoam -n redhat-rhoam-operator -o yaml | yq - r 'status.sku'` -> 20

### Verify that the 3scale has expected values after SKU has changed

#### Memory and cpu 
Apicast production: 
```
oc get apimanager 3scale -n redhat-rhoam-3scale -o yaml | yq - r 'spec.apicast.productionSpec.resources'
```

Output:
```
limits:
  cpu: 600m
  memory: 300Mi
requests:
  cpu: 300m
  memory: 250Mi
```
Backend listeners: 
```
oc get apimanager 3scale -n redhat-rhoam-3scale -o yaml | yq - r 'spec.backend.listenerSpec.resources'
```
Output: 
```
limits:
  cpu: 600m
  memory: 500Mi
requests:
  cpu: 250m
  memory: 450Mi
```

Backend workers: 
```
oc get apimanager 3scale -n redhat-rhoam-3scale -o yaml | yq - r 'spec.backend.workerSpec.resources'
```

Output: 
```
limits:
  cpu: 300m
  memory: 100Mi
requests:
  cpu: 150m
  memory: 100Mi
```

#### Replicas 
Apicast production:

`oc get apimanager 3scale -n redhat-rhoam-3scale -o yaml | yq - r 'spec.apicast.productionSpec.replicas'`-> 3

Backend listeners: 

`oc get apimanager 3scale -n redhat-rhoam-3scale -o yaml | yq - r 'spec.backend.listenerSpec.replicas' ` -> 3

Backend workers: 

`oc get apimanager 3scale -n redhat-rhoam-3scale -o yaml | yq - r 'spec.backend.workerSpec.replicas'` -> 3


### Verify that manual changes to resources and replicas will re-update outside of skuupdate event when the value is lower
Patch the replicas of any of: apicast production, backend worker or backend listener; in the `APIManager` CR to 1. This can be done via the UI.
Patch any of the limits and and requests below the most previous values
e.g. for backend workers:
```
limits:
  cpu: 199m
  memory: 0Mi
requests:
  cpu: 2m
  memory: 60Mi
```
Wait ~5 minutes for reconcile to complete

Verify that the operator has put the value back to expected.

Comand end expected values from the verification spet before. 
e.g for the Apicast production: 
```
oc get apimanager 3scale -n redhat-rhoam-3scale -o yaml | yq - r 'spec.apicast.productionSpec.resources'
```

Output:
```
limits:
  cpu: 600m
  memory: 300Mi
requests:
  cpu: 300m
  memory: 250Mi
```

### Verify that lowering the sku works as expected
Run `INSTALLATION_TYPE=managed-api QUOTA=1 make cluster/prepare/sku`

which will update the secret with a sku value of 1

`oc get rhmi rhoam -n redhat-rhoam-operator -o yaml | yq - r 'status.toSKU'` -> 1

Wait for the reconcile to complete

`oc get rhmi rhoam -n redhat-rhoam-operator -o yaml | yq - r 'status.toSKU'` -> nil
`oc get rhmi rhoam -n redhat-rhoam-operator -o yaml | yq - r 'status.sku'` -> 1

Verify the values of the APIManager are as expected

By running the commands from the beginning `sku==1`
e.g for Apicast production: 
```oc get apimanager 3scale -n redhat-rhoam-3scale -o yaml | yq - r 'spec.apicast.productionSpec.resources'```

Output:
```
limits:
  cpu: 100m
  memory: 270Mi
requests:
  cpu: 90m
  memory: 250Mi
```

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist
- [ ] This change requires a documentation update
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added a test case that will be used to verify my changes 
- [ ] Verified independently on a cluster by reviewer

makslion commented:
It is expected for e2e test to fails. Failing tests are being fixed by (this)[https://issues.redhat.com/browse/MGDAPI-1488] JIRA

laurafitzgerald commented on pkg/products/threescale/reconciler.go:
```diff
@@ -1043,26 +1055,35 @@ func (r *Reconciler) reconcileComponents(ctx context.Context, serverClient k8scl
 
 		owner.AddIntegreatlyOwnerAnnotations(apim, r.installation)
 
+		if err != nil {
```
should this error check be moved up to inside the if statement for the Configure function? Line 1079

makslion replied:
moved inside the if statement

laurafitzgerald commented on pkg/resources/sku/sku.go:
```diff
@@ -28,7 +29,20 @@ const (
 
 var (
-	defaultReplicas int32 = 1
+	products = map[v1alpha1.ProductName][]string{
```
This is defined twice in this file. Once copy can be removed.

makslion replied:
moved

makslion commented:
changes with the last commit: 
- removed apicast staging from SKU configuration 
- removed duplicate 3scale import to the scheme in sku_test.go
- removed duplicate of `products` map in the sku.go 
- united if statements to skip value assignment In case of RHOAM to values being configured by the SKU

MStokluska commented on pkg/resources/sku/sku.go:
```diff
@@ -190,28 +182,60 @@ func (p AProductConfig) Configure(obj metav1.Object) error {
+
+func (p AProductConfig) mutateReplicas(replicas *int32, name string) {
+	if replicas == nil {
+		replicas = &[]int32{0}[0]
```
@sergioifg94 I think I could use help with reviewing this bit of code and wonder if you could share your opinion on it?

sergioifg94 replied:
There's a couple of things I'm not sure about in this function.

1. When the value of `replicas` is `nil`, we're only assigning the local variable in the function, but not the value that is passed to it, so if we try the following, the value of `myReplicas` will still be `nil`:
    ```go
    var myReplicas *int32 = nil
   productConfig.mutateReplicas(myReplicas, name)

   fmt.Println(myReplicas) // prints <nil>
    ```

    Is there any test that verifies this? If it's ok to leave the value unchanged when it's `nil`, it'd be better to simply `return` in the `if` block
2. When assigning the value of `replicas`, Is there a reason why we take the address of the first element of a newly created slice instead of using `new(int32)`?

MStokluska commented:
/lgtm

Verified it on the cluster with RHOAM installation, initial SKU, and update to SKU works as expected with the values listed in the verification steps. 

Code was verified as well however, due to the number of questions I had it was done on gchat instead. 

Error was caught during scaling of SKU from 1 to 20, however, this is not related to the work in this PR and the ticket for addressing it can be found here - https://issues.redhat.com/browse/MGDAPI-1592 

@makslion @laurafitzgerald thanks a lot for all your explanations. Good job!
